File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double _real;
    private readonly double _imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        _real = real;
        _imaginary = imaginary;
    }

    public double Real() => _real;

    public double Imaginary() => _imaginary;

    public ComplexNumber Mul(ComplexNumber other) => new(Real() * other.Real() - Imaginary() * other.Imaginary(), Imaginary() * other.Real() + Real() * other.Imaginary());

    public ComplexNumber Mul(int other) => new(Real() * other, Imaginary() * other);

    public ComplexNumber Add(ComplexNumber other) => new(Real() + other.Real(), Imaginary() + other.Imaginary());

    public ComplexNumber Add(int other) => new(Real() + other, Imaginary());

    public ComplexNumber Sub(ComplexNumber other) => new(Real() - other.Real(), Imaginary() - other.Imaginary());

    public ComplexNumber Div(ComplexNumber other)
    {
        double real = (Real() * other.Real() + Imaginary() * other.Imaginary()) / (other.Real() * other.Real() + other.Imaginary() * other.Imaginary());
        double imaginary = (Imaginary() * other.Real() - Real() * other.Imaginary()) /
                        (Math.Pow(other.Real(), 2) + Math.Pow(other.Imaginary(), 2));
        return new(real, imaginary);
    }

    public ComplexNumber Div(int other) => new(Real() / other, Imaginary() / other);

    public double Abs() => Math.Sqrt(Math.Pow(Real(), 2) + Math.Pow(Imaginary(), 2));

    public ComplexNumber Conjugate() => new(Real(), -Imaginary());

    public ComplexNumber Exp()
    {
        double first = Math.Pow(Math.E, Real());
        return new(first * Math.Cos(Imaginary()), first * Math.Sin(Imaginary()));
    }
}

File: cpp/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"
#include <cmath>

namespace complex_numbers
{

    Complex::Complex(double real, double imag) : _real(real), _imag(imag) {}

    double Complex::real() const
    {
        return _real;
    }

    double Complex::imag() const
    {
        return _imag;
    }

    Complex Complex::operator+(const Complex &other) const
    {
        return Complex{_real + other._real, _imag + other._imag};
    }

    Complex Complex::operator+(double other) const
    {
        return Complex{_real + other, _imag};
    }

    Complex Complex::operator-(const Complex &other) const
    {
        return Complex{_real - other._real, _imag - other._imag};
    }

    Complex Complex::operator-(double other) const
    {
        return Complex{_real - other, _imag};
    }

    Complex Complex::operator*(const Complex &other) const
    {
        return Complex{_real * other._real - _imag * other._imag, _imag * other._real + _real * other._imag};
    }

    Complex Complex::operator*(double other) const
    {
        return Complex{_real * other, _imag * other};
    }

    Complex Complex::operator/(const Complex &other) const
    {
        double denom = (pow(other._real, 2.0) + pow(other._imag, 2.0));
        double real = (_real * other._real + _imag * other._imag) / denom;
        double imag = (_imag * other._real - _real * other._imag) / denom;
        return Complex{real, imag};
    }

    Complex Complex::operator/(double other) const
    {
        return Complex{_real / other, _imag / other};
    }

    Complex operator+(double lhs, const Complex &rhs)
    {
        return Complex{lhs + rhs._real, rhs._imag};
    }

    Complex operator-(double lhs, const Complex &rhs)
    {
        return Complex{lhs - rhs._real, -rhs._imag};
    }

    Complex operator*(double lhs, const Complex &rhs)
    {
        return Complex{lhs * rhs._real, lhs * rhs._imag};
    }

    Complex operator/(double lhs, const Complex &rhs)
    {
        double denom = (pow(rhs._real, 2.0) + pow(rhs._imag, 2.0));
        return Complex{lhs * rhs._real / denom, -lhs * rhs._imag / denom};
    }

    double Complex::abs() const
    {
        return sqrt(pow(_real, 2.0) + pow(_imag, 2));
    }

    Complex Complex::exp() const
    {
        double first = pow(std::exp(1.0), _real);

        return Complex{first * cos(_imag), first * sin(_imag)};
    }

    Complex Complex::conj() const
    {
        return Complex{_real, -_imag};
    }

} // namespace complex_numbers
